@here/harp-utils/lib/TaskQueue.ts

```diff
@@ -0,0 +1,235 @@
+     */
+    isExpired?: () => boolean;
+
+    /**
+     * An optional function that returns an estimated process time,
+     * this is not directly used by the [[TaskQueue]] but can be used
+     * by an Task Scheduler to schedule the processing
+     */
+    estimatedProcessTime?: () => number;
+}
+
+export interface TaskQueueOptions {
+    //TODO: Use the max length
+    //maxLength?: number;
+    /**
+     * Groups to combine specific [[Task]]s in the [[TaskQueue]],
+     * [[Task]]s can only be added to the [[TaskQueue]] if their group is available
+     */
+    groups: string[];
+
+    /**
+     * Optional function to sort the priority, if set, i will override the internal TaskQueue.sort
+     * function.
+     * Caution, the [[TaskQueue]] uses the last element in the Arrays first, so the
+     * highest priorities should be ordered to the end
+     */
+    prioSortFn?: (a: Task, b: Task) => number;
+}
+
+/**
+ * This is a helper interface to be able to index the TaskArrays via a string
+ */
+interface TaskArrays {
+    [key: string]: Task[];
+}
+
+/**
+ * A Pull-TaskQueue sorted by priority and group-able [[Task]]s by [[Task.group]]
+ *
+ * Sample Usage
+ * ```
+ *  const taskQueue = new TaskQueue({
+ *      group: ["group1"]
+ *  })
+ *  taskQueue.add({
+ *     group: "group1",
+ *     execute: () => {
+ *         console.log("task of group1 executed");
+ *     },
+ *     getPrio: () => {
+ *         return 0;
+ *     }
+ *   });
+ *
+ * taskQueue.update();
+ * taskQueue.processNext("group1");
+ *
+ *  ```
+ */
+export class TaskQueue {
+    private m_taskArrays: TaskArrays = {};
+
+    constructor(private m_options: TaskQueueOptions) {
+        this.m_options.groups?.forEach(group => {
+            this.m_taskArrays[group] = [];
+        });
+        this.sort = this.m_options.prioSortFn ?? this.sort;
+    }
+
+    /**
+     * Updates the lists in the queue depending on their priority functions
+     * Needs to be called by using instance, as it would be to costly on every change
+     * @param group The Group to update, if not set all groups will be updated.
+     */
+    update(group?: string) {
+        if (group === undefined) {
+            this.forAllGroups(taskList => {
+                this.updateTaskList(taskList);
+            });
+        } else {
+            const taskList = this.getTaskList(group);
+            if (taskList) {
+                this.updateTaskList(taskList);
+            }
+        }
+    }
+
+    /**
+     * Adds a Task to the Queue
+     *
+     * @param task
+     * @returns true if succesfully added, otherwise false
+     */
+    add(task: Task): boolean {
+        if (this.m_taskArrays[task.group]) {
+            this.m_taskArrays[task.group].push(task);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Removes a Task to the Queue
+     *
+     * @param task
+     * @returns true if succesfully removed, otherwise false
+     */
+    remove(task: Task): boolean {
```

Comment:
Why do we have this method? Don't we always pop the last one (i.e. the one with the highest priority)?

Reply:
yes, but you can remove the Task from the TaskQueue if you decide it should not be processed anymore